### ISSUE TITLE
Unable to apply | Logitech MX Master 3

### ISSUE THREAD

**Issue by RAYs3T** [libratbag/piper#433]
Piper Version: 0.3
Device: Logitech MX Master 3

Error when applying settings.

**Comment by Dannyzen** [comment]
Jumping in (slickdeals had a recent $25 coupon on this device and I couldn't hold back) 
https://github.com/libratbag/libratbag/pull/819 shows a valid pull. 
I run & get
```
$ratbagd
> hidraw1: "Logitech MX Master 3000", 1 profiles
```

But Piper does not display the device nor can I change or apply any settings.
![image](https://user-images.githubusercontent.com/659908/70198622-56ea4a80-16dd-11ea-9b59-dee70e078eb1.png)


Any thoughts?

**Comment by jebaum** [comment]
Chiming in to say I'm seeing the same exact thing. I can also issue commands through ratbagctl just fine. Some system information:

Linux 5.3.0-24-generic #26~18.04.2-Ubuntu SMP Tue Nov 26 12:34:22 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=18.04
DISTRIB_CODENAME=bionic
DISTRIB_DESCRIPTION="Pop!_OS 18.04 LTS"

ratbagctl --version -> 0.11
ratbagd --version -> 0.11

**Comment by whot** [comment]
Try kernel 5.4 if you can please. IIRC a bunch of changes went into that kernel and this mouse was one of them.

**Comment by FFY00** [comment]
That's not the issue, as I mentioned in https://github.com/libratbag/piper/issues/442 `0x1b04` is not working.

**Comment by Dannyzen** [comment]
Friendly bump. 
Notable: If a 'donation' of this mouse could help escalate this issue's priority (and the mouses' support), I would be happy to assist.

**Comment by FFY00** [comment]
Fortunately a donation is not needed, I have a sample. The issue right now is finding time to dive into the code and figure out what's going wrong here.

If you have a little bit of time, one thing you could do is to see if this was working in previous versions (it should) and bisect the breaking commit.

**Comment by jebaum** [comment]
I built from ca4a1e9 as well as 172d34e and got the same 404 error. Any build prior to 172d34e fails with the following (which seems expected based on the content of commit 172d34e):

```
Traceback (most recent call last):
  File "./builddir/piper", line 46, in <module>
    exit_status = Application().run(sys.argv)
TypeError: __init__() missing 1 required positional argument: 'ratbagd_api_version'
```

Note that all I did was follow the exact steps listed in the readme of this repo, namely:

```
$ git clone https://github.com/libratbag/piper.git
$ cd piper
$ meson builddir --prefix=/usr/
$ ninja -C builddir
```

Then ran the resulting binary that gets created in `./builddir`

**Comment by rscircus** [comment]
+1

**Comment by numman-ali** [comment]
+1 
Happening on POP_OS 19.10, latest version of libratbag and piper

**Comment by leirbagbrito** [comment]
Good evening guys, any news about this? I am using Fedora 31 and it is giving this same error.

**Comment by FFY00** [comment]
Yes, https://github.com/libratbag/libratbag/pull/890 should fix it.

**Comment by leirbagbrito** [comment]
When will version 0.12 be available in the Fedora 31 repository?

**Comment by FFY00** [comment]
I am not sure. But the fix will only be included in 0.13.

**Comment by dionearle** [comment]
I am still having these issues with my MX Master 3 on Ubuntu, is there a current fix available? Thanks

**Comment by FFY00** [comment]
It is fixed in master. Currently waiting on #904 for a new release.

**Comment by dpwrussell** [comment]
I am still seeing this problem with libratbag 0.13 I'm afraid.

```
$ apt search ratbag
Sorting... Done
Full Text Search... Done
libratbag-tools/eoan,eoan,now 0.13-1+git1594-02c9542~ubuntu19.10.1 all [installed,automatic]
  configuration tools for gaming mice (transitional package)

piper/eoan,eoan,now 0.4+git429-06ac2ee~ubuntu19.10.1 all [installed]
  graphical frontend for libratbag

ratbagd/eoan,now 0.13-1+git1594-02c9542~ubuntu19.10.1 amd64 [installed,automatic]
  D-Bus daemon handling mouse configuration
```

**Comment by FFY00** [comment]
Start the daemon with verbosity

```
ratbagd --verbose=raw
```

And paste the log when trying to apply the changes.

**Comment by thor27** [comment]
Building from master I was able to make it work but not fully. 

I can't map to macro, the default map for the "thumb" button is marked as UNKNOWN (it seems to be SUPER+TAB). 

I would love to map this button to take screenshot as I do on Mac OS X.

I tried to map to button (button 6) for example but it doesn't work also.

All times it gave an error on piper the message I receive from `ratbagd --verbose` is:

```
ratbag error:     HID++ error from the device (3): ERR_INVALID_ARGUMENT (02)
ratbag error: Error while writing profile: 'Protocol error' (-71)
ratbag error: hidpp20: failed to update button (-71)
ratbagd error: error committing device (-1000)
```

Anyway, it worked to map the "thumb" button to back, front and middle click.

**Comment by mielouk** [comment]
I can confirm that it is mapped to Super+Tab by default. This combination has been mapped to vertical split in my WM and the "thumb" button does select the vertical split.

**Comment by bashfulrobot** [comment]
I am just piping in here to offer more info if needed as I am experiencing the same issue. If you need testing or queries from the system - let me know. 

Thank you.

## System

### Distro

```
❯ cat /etc/lsb-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=20.04
DISTRIB_CODENAME=focal
DISTRIB_DESCRIPTION="Ubuntu 20.04 LTS"
```

### Ratbag

```
ratbagd/focal,now 0.13-1 amd64
```

### Piper

```
piper/focal,focal,now 0.4-1 all
```

### MX Master 3

```
❯ ratbagctl
hollering-marmot:    Logitech MX Master 3

❯ solaar show mouse
  2: MX Master 3
     Codename     : MX Master 3
     Kind         : mouse
     Wireless PID : 4082
     Protocol     : HID++ 4.5
     Polling rate : 8 ms (125Hz)
     Serial number: xxxxxxxxxx
     The power switch is located on the base.
     Battery status unavailable.
```

**Comment by FFY00** [comment]
@thor27 sorry for the delay, please open a new issue for that. We should close this.

@bashfulrobot what exactly are you experiencing? Failing to apply? In that case, have you tried master?

**Comment by harshitgoel96** [comment]
Hi. It's still an issue. I see 404 instead of the mouse, and apply does nothing

**Comment by Cthululz** [comment]
Same here

**Comment by Hamza-Anver** [comment]
Since this isn't closed would like to note that the MX Master 3 works for me with piper in 2023 on Fedora running piper 0.7

**Comment by thor27** [comment]
Just tested, latest version from git, fresh ubuntu install, still not working.

Mouse detects fine, but can't assign anything to the thumb button, same error as before.

**Comment by theniwo** [comment]
Since this isn't closed I would like to bump, that my G502 X doesn't work with either ubuntu 22.04 nor Fedora 37.
It gives the 404 Image and buttons are configurable, but not working.

```
ratbagctl list 
warbling-mara:       Logitech G502 X 
lsusb | grep 502
Bus 001 Device 040: ID 046d:c099 Logitech, Inc. G502 X
```

**Versions**
Fedora
```
6.2.7-200.fc37.x86_64
libratbag-ratbagd-0.16-4.fc37.x86_64
piper-0.7-2.fc37.noarch

```
I also tried 
libratbag-ratbagd-0.17-2.fc38.x86_64
piper-0.7-3.fc38.noarch

Ubuntu 
tbd

**Comment by Roccoriu** [comment]
I'm running Fedora 39 and the issue still is persistent. 

```
Oct 09 09:11:33 fedora systemd[1]: Starting ratbagd.service - Daemon to introspect and modify configurable mice...
Oct 09 09:11:33 fedora systemd[1]: Started ratbagd.service - Daemon to introspect and modify configurable mice.
Oct 09 09:11:41 fedora ratbagd[3250]: ratbag error: hidpp20: failed to update report rate (-95)
Oct 09 09:11:41 fedora ratbagd[3250]: ratbagd error: error committing device (-1000)
Oct 09 09:11:45 fedora ratbagd[3250]: ratbag error: hidpp20: failed to update report rate (-95)
Oct 09 09:11:45 fedora ratbagd[3250]: ratbagd error: error committing device (-1000)
Oct 09 09:12:01 fedora ratbagd[3250]: ratbag error: hidpp20: failed to update report rate (-95)
Oct 09 09:12:01 fedora ratbagd[3250]: ratbagd error: error committing device (-1000)
Oct 09 09:12:03 fedora ratbagd[3250]: ratbag error: hidpp20: failed to update report rate (-95)
Oct 09 09:12:03 fedora ratbagd[3250]: ratbagd error: error committing device (-1000)
```

I'm running the following system:
Kernel: 6.5.6-300.fc39.x86_64
piper: 0.7
ratbagctl: 0.17
Mouse: Mx master 3s

**Comment by aria-dos** [comment]
Similar issue with Logitech MX Vertical on Debian (piper version 0.7-1)

**Comment by nothingneko** [comment]
still having this issue in good 'ol 2025

![Image](https://github.com/user-attachments/assets/2e74857b-d021-4bbd-900b-ad1598f59bc0)

![Image](https://github.com/user-attachments/assets/631bb42a-a4ff-4f71-95d8-e5791c10f539)

![Image](https://github.com/user-attachments/assets/b859dd64-f29c-4c48-8238-ac093771750e)
